pkg/pillar/cmd/zedagent/handlemetrics.go

```diff
@@ -528,6 +531,50 @@ func publishMetrics(ctx *zedagentContext, iteration int) {
 	}
 	log.Tracef("DirPaths in persist, elapse sec %v", time.Since(startPubTime).Seconds())
 
+	// ZFS metrics
+	if vault.ReadPersistType() == types.PersistZFS {
+		zpoolList, err := libzfs.PoolOpenAll()
+		if err != nil {
+			log.Errorf("get zpool list failed %v", err)
+		} else {
+			for _, zpool := range zpoolList {
+				rStorageMetric := new(zmet.StorageMetric)
+				rStorageMetric.PoolName = zpool.Properties[libzfs.PoolPropName].Value
+				rStorageMetric.StorageState = zfs.GetZfsDeviceStatusFromStr(zpool.Properties[libzfs.PoolPropHealth].Value)
+				rStorageMetric.CountVolume, err = zfs.GetZfsCountVolume(zpool.Properties[libzfs.PoolPropName].Value)
```

Comment:
This logic do not work in my tests. We should call `GetZfsCountVolume` on `types.VolumeClearZFSDataset` and `types.VolumeEncryptedZFSDataset` and sum results OR we should rework `GetZfsCountVolume` to go into nested datasets.

Reply:
Corrected

Reply:
Now I can see that we will handle the second level of children, but we expect the third (something like persist/vault/volumes/zvol1)

Reply:
Can we try to get rid of libzfs.DatasetOpen(zpoolName)? I understand that we have same values for pool and dataset, but naming of variables and comments of functions looks odd.

Reply:
Corrected. The path won't matter now.